### PR TITLE
Experimental support for PowerShell 7 and Linux (for non-build jobs)

### PR DESCRIPTION
**Pull request by freddydk**
### Experimental Support
Setting the repo setting "shell" to "pwsh", followed by running Update AL-Go System Files, will cause all PowerShell code to be run using PowerShell 7 instead of PowerShell 5. This functionality is experimental. Please report any issues at https://github.com/microsoft/AL-Go/issues

Setting the repo setting "runs-on" to "Ubuntu-Latest", followed by running Update AL-Go System Files, will cause all non-build jobs to run using Linux. This functionality is experimental. Please report any issues at https://github.com/microsoft/AL-Go/issues

### CODE REVIEW

**Comment by freddydk** [comment]
Running end-2-end tests with these changes here: https://github.com/microsoft/AL-Go/actions/runs/3682647177

File: Actions/AL-Go-Helper.ps1 [hunk 1]
```diff
@@ -469,13 +482,18 @@ function ReadSettings {
...
-                throw "Settings file $settingsFile, is wrongly formatted. Error is $($_.Exception.Message)."
+                throw "Settings file $settingsFile, is wrongly formatted. Error is $($_.Exception.Message).`n$($_.ScriptStackTrace)"
             }
         }
     }
 
     if ($settings.githubRunner -eq "") {
-        $settings.githubRunner = $settings."runs-on"
+        if ($settings."runs-on" -like "ubuntu-*") {
+            $settings.githubRunner = "windows-latest"
```
Comment (mazhelez) [inline Actions/AL-Go-Helper.ps1:492]: The other way around?

File: Actions/AL-Go-TestRepoHelper.ps1 [hunk 2]
```diff
@@ -64,12 +64,14 @@ function Test-ALGoRepository {
...
+    Write-Host "BaseFolder: $baseFolder"
+
     # Test .json files are formatted correctly
     Get-ChildItem -Path $baseFolder -Filter '*.json' -Recurse | ForEach-Object {
-        if ($_.FullName -like '*\.AL-Go\Settings.json') {
+        if ($_.FullName -like "*$([System.IO.Path]::DirectorySeparatorChar).AL-Go$([System.IO.Path]::DirectorySeparatorChar)settings.json") {
             Test-Json -jsonFile $_.FullName -baseFolder $baseFolder
         }
-        elseif ($_.FullName -like '*\.github\*Settings.json') {
+        elseif ($_.FullName -like "*$([System.IO.Path]::DirectorySeparatorChar).github$([System.IO.Path]::DirectorySeparatorChar)*ettings.json") {
```
Comment (mazhelez) [inline Actions/AL-Go-TestRepoHelper.ps1:74]:
```suggestion
        elseif ($_.FullName -like "*$([System.IO.Path]::DirectorySeparatorChar).github$([System.IO.Path]::DirectorySeparatorChar)settings.json") {
```

Maybe? Hopefully?

**Review by mazhelez — COMMENTED** [review]
Consider splitting some scripts into smaller chunks that could be tested separately. A starter candidate is CheckForUpdates.ps1

File: Actions/ReadSettings/ReadSettings.ps1 [hunk 3]
```diff
@@ -101,14 +101,16 @@ try {
...
+        Write-Host "Determining projects to build"
 
         $buildProjects = @()
-        if ($settings.Projects) {
+        if ($settings.projects) {
             $projects = $settings.projects
         }
         else {
-            $projects = @(Get-ChildItem -Path $ENV:GITHUB_WORKSPACE -Directory -Recurse -Depth 2 | Where-Object { Test-Path (Join-Path $_.FullName '.AL-Go\Settings.json') -PathType Leaf } | ForEach-Object { $_.FullName.Substring("$ENV:GITHUB_WORKSPACE".length+1) })
+            $projects = @(Get-ChildItem -Path $ENV:GITHUB_WORKSPACE -Recurse -Depth 2 | Where-Object { $_.PSIsContainer -and (Test-Path (Join-Path $_.FullName ".AL-Go$([System.IO.Path]::DirectorySeparatorChar)settings.json") -PathType Leaf) } | ForEach-Object { $_.FullName.Substring("$ENV:GITHUB_WORKSPACE".length+1) })
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/ReadSettings/ReadSettings.ps1:113]: I know, but the code needs to be able to run on PS5 as well - lowest common denominator decides.

File: Actions/AL-Go-Helper.ps1 [hunk 4]
```diff
@@ -10,9 +10,9 @@ if (Test-Path $gitHubHelperPath) {
 $ErrorActionPreference = "stop"
 Set-StrictMode -Version 2.0
 
-$ALGoFolder = ".AL-Go\"
-$ALGoSettingsFile = ".AL-Go\settings.json"
-$RepoSettingsFile = ".github\AL-Go-Settings.json"
+$ALGoFolder = ".AL-Go$([System.IO.Path]::DirectorySeparatorChar)"
+$ALGoSettingsFile = ".AL-Go$([System.IO.Path]::DirectorySeparatorChar)settings.json"
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:14]: Good point, will change

File: Actions/AL-Go-Helper.ps1 (hunk 1 above)
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:492]: No,

"runs-on" is the runner used for all jobs (except for build jobs)
GitHubRunner is the runner used for build jobs (default is to use runs-on)
Build jobs doesn't support Ubuntu.

File: Actions/AL-Go-TestRepoHelper.ps1 (hunk 2 above)
Comment (freddydk) [reply Actions/AL-Go-TestRepoHelper.ps1:74]: No, This is reading all settings files in .github folder (and tests validity):
image

File: Actions/AL-Go-Helper.ps1 [hunk 5]
```diff
@@ -204,7 +217,7 @@ function DownloadAndImportBcContainerHelper {
             if (-not $BcContainerHelperVersion) {
                 $repoSettings = Get-Content $repoSettingsPath -Encoding UTF8 | ConvertFrom-Json | ConvertTo-HashTable
                 Write-Host $ENV:GITHUB_ACTION_PATH
-                $ap = "$ENV:GITHUB_ACTION_PATH".Split('\')
+                $ap = "$ENV:GITHUB_ACTION_PATH".Split([System.IO.Path]::DirectorySeparatorChar)
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:220]: will do:-)

**Comment by freddydk** [comment]
End-2-End tests running Ubuntu here: https://github.com/microsoft/AL-Go/actions/runs/3684994328

File: Actions/ReadSettings/ReadSettings.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/ReadSettings/ReadSettings.ps1:117]: On splitting - I think it is a good point, and something I haven't been good at.
AL-Go has a very good coverage in end-2-end testing - but it isn't sufficient.
I have an item called Test Dept, and the plan is to spend a few sprints to catch up on CI tests more broadly.

File: Actions/ReadSettings/ReadSettings.ps1 [hunk 6]
```diff
@@ -150,7 +156,7 @@ try {
                     $buildProjects = @($projects | Where-Object {
                         $project = $_
                         $buildProject = $false
-                        if (Test-Path -path (Join-Path $ENV:GITHUB_WORKSPACE "$project\.AL-Go\Settings.json")) {
+                        if (Test-Path -path (Join-Path $ENV:GITHUB_WORKSPACE "$project$([System.IO.Path]::DirectorySeparatorChar).AL-Go$([System.IO.Path]::DirectorySeparatorChar)settings.json")) {
```
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:159]: Are you sure the `$([System.IO.Path]::DirectorySeparatorChar).AL-Go$([System.IO.Path]::DirectorySeparatorChar)` is needed?

PS seems to be pretty smart about the slashes:
```
Join-Path 'a' 'b/c\d/e'
a\b\c\d\e
```

**Review by mazhelez — COMMENTED** [review]
Left some comments here and there. I think there could be improvements. The usage of `$([System.IO.Path]::DirectorySeparatorChar)` is a thorn in the side: it decreases the readability :/
My suggestion is to use `Join-Path` whenever possible.

File: Actions/AL-Go-Helper.ps1 (hunk 1 above)
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:499]: AL-Go doesn't support macos.
Partners will set runs-on to self-hosted in order to use the self-hosted for all jobs.
It is only if they set ubuntu that we won't use ubuntu for GitHubRunner (which is the build job runner)

File: Actions/AL-Go-Helper.ps1 [hunk 7]
```diff
@@ -544,7 +572,7 @@ function AnalyzeRepo {
     }
 
     if (-not (@($settings.appFolders)+@($settings.testFolders)+@($settings.bcptTestFolders))) {
-        Get-ChildItem -Path $projectPath -Directory | Where-Object { Test-Path -Path (Join-Path $_.FullName "app.json") } | ForEach-Object {
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:547]: When using -Directory on Linux I got a different object type back.
It seemed like Get-ChildItem with Directory was mapped to another linux construct and I couldn't get the FullName from the object.

File: Actions/CheckForUpdates/CheckForUpdates.ps1 [hunk 8]
```diff
@@ -84,185 +84,191 @@ try {
...
+                            $newBuild = @()
+                            $build = $yaml.Get('jobs:/Build:/')
+                            1..$depth | ForEach-Object {
+                                if ($_ -eq 1) {
+                                    $needs = @('Initialization')
+                                    $if = "if: needs.Initialization.outputs.projects$($_)Count > 0"
+                                }
+                                else {
+                                    $newBuild += @('')
+                                    $needs = @('Initialization',"Build$($_-1)")
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/CheckForUpdates/CheckForUpdates.ps1:232]: This is only determining the build order.
The subsequent job download the artifacts from all previous builds.

File: Actions/ReadSettings/ReadSettings.ps1 (hunk 6 above)
Comment (freddydk) [reply Actions/ReadSettings/ReadSettings.ps1]: I didn't know that - I have been through all occurances of DirectorySeperatorChar and modified to Join-Path

File: Actions/AL-Go-Helper.ps1 [hunk 9]
```diff
@@ -204,7 +217,13 @@ function DownloadAndImportBcContainerHelper {
...
                 $repoSettings = Get-Content $repoSettingsPath -Encoding UTF8 | ConvertFrom-Json | ConvertTo-HashTable
                 Write-Host $ENV:GITHUB_ACTION_PATH
-                $ap = "$ENV:GITHUB_ACTION_PATH".Split('\')
+                $ap = "$ENV:GITHUB_ACTION_PATH".Split([System.IO.Path]::DirectorySeparatorChar)
+                # Action Path under linux is something like: /home/runner/work/_actions/freddydk/AL-Go-Actions/main/WorkflowInitialize
+                # Action Path under Windows is something like: D:\a\_actions\freddydk\AL-Go-Actions\main\WorkflowInitialize
+                # ..../_actions/<owner>/AL-Go-Actions/branch/Action
+                # ..../   -5   /  -4   /     -3      /  -2  /  -1
+                # The code below checks if you are running a preview branch of AL-Go-Actions - then use Preview ContainerHelper
+                # If you are using a private fork of AL-Go-Actions - then use (if exists) a private fork of ContainerHelper as well
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:226]: I have blocked the setting from being able to point to a URL. Allowing people create create a PR containing a URL in a settings file to a module, which would be downloaded and run in context with access to secrets and more seems wrong.

I don't like that people would need to update system files and subsequently modify the settings file to point to a private containerhelper in order to work with development of AL-Go and ContainerHelper.
and... - this only happens for people actually developing on AL-Go - this wouldn't accidently happen for any partners.

But... - maybe I should setup a dependency version number - stating that this AL-Go repo requires BcContainerHelper version x.x or higher - and then fail on wrong settings - or stale forks.

File: Actions/AL-Go-Helper.ps1 [hunk 10]
```diff
@@ -327,7 +346,7 @@ function MergeCustomObjectIntoOrderedDictionary {
             if ($srcPropType -eq "PSCustomObject" -and $dstPropType -eq "OrderedDictionary") {
                 MergeCustomObjectIntoOrderedDictionary -dst $dst."$prop" -src $srcProp
             }
-            elseif ($dstPropType -ne $srcPropType) {
+            elseif ($dstPropType -ne $srcPropType -and !($srcPropType -eq "Int64" -and $dstPropType -eq "Int32")) {
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:349]: dstPropType is always Int32 (comes from code)
On lInux, reading a json file makes int's become Int64

File: Actions/AL-Go-Helper.ps1 [hunk 11]
```diff
@@ -425,14 +443,16 @@ function ReadSettings {
         "appDependencyProbingPaths"              = @()
         "useProjectDependencies"                 = $false
         "runs-on"                                = "windows-latest"
+        "shell"                                  = "powershell"
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:446]: The problem of updating the documentation in the PR is that the documentation will be updated in the main branch immediately.
Need to find a good way to have preview documentaion - we don't have that right now.

File: Actions/AL-Go-Helper.ps1 (hunk 9 above)
[the remark this answers is not in the document]
Comment (mazhelez) [reply Actions/AL-Go-Helper.ps1:226]: Working on a old version of _ContainerHelper_ is what actually happened earlier this week.

Imagine someone contributed a feature to the ContainerHelper two years ago and now they want to contribute to AL-Go. The logic would silently fallback to their 2-year old version.
There needs to be an explicit setting where to take the ContainerHelper from.

File: Actions/AL-Go-Helper.ps1 (hunk 1 above)
Comment (mazhelez) [reply Actions/AL-Go-Helper.ps1:499]: If _runs-on_ is _self-hosted_, then _self-hosted_ is used for all jobs.
If _runs-on_ is _ubuntu-*_ then, the build jobs will be on _windows-latest_ and the non-build jobs: on _ubuntu-*_.

Aren't these fallback settings hard to follow and maintain?

Hear me out: _githubrunner_ defaults to _runs-on_, _runs_on_ defaults to _windows-latest_. Period.
Worst case, partners have to set up one more setting.

Imagine in X time from now, if we can actually build BC on Linux, how this fallback won't make sense.

Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:499]: If runs-on is self-hosted, then self-hosted is used for all jobs.
If runs-on is ubuntu-* then we will use ubuntu for all the jobs possible (and windows if the job cannot run Linux)

Re: 
_Imagine in X time from now, if we can actually build BC on Linux, how this fallback won't make sense._

Which is exactly the reason why they should NOT set gitHubRunner to Windows-latest.

If they set runs-on to Ubuntu - we will run everything (possible) on Ubuntu.
and in x months from now where we can run builds on Linux - we will do that.
If everybody needs to add a setting to set GitHubRunner Windows-Latest, then they will have to change settings again when we support Linux for builds as well.

File: Actions/AL-Go-TestRepoHelper.ps1 (hunk 2 above)
Comment (mazhelez) [reply Actions/AL-Go-TestRepoHelper.ps1]: Wait, I do not get it... why isn't `*settings.json"` matching all settings files?

Comment (freddydk) [reply Actions/AL-Go-TestRepoHelper.ps1]: Linux has case sensitive file names - and we have:
.AL-Go\settings.json (project settings)
.github\AL-Go-Settings.json (repo settings)
.github\CICD.settings.json (workflow settings)

The code after removing the directory separators is:

```
    # Test .json files are formatted correctly
    Get-ChildItem -Path $baseFolder -Filter '*.json' -Recurse | ForEach-Object {
        if ($_.FullName -like (Join-Path "*/.AL-Go" "settings.json")) {
            Test-Json -jsonFile $_.FullName -baseFolder $baseFolder
        }
        elseif ($_.FullName -like (Join-Path "*/.github" "*ettings.json")) {
            Test-Json -jsonFile $_.FullName -baseFolder $baseFolder -repo:($_.BaseName -eq "AL-Go-Settings")
        }
    }
```
I guess this would work as well:

```
Get-ChildItem -Path $baseFolder -Filter '*ettings.json' -Recurse -File | Where-Object { $_.Directory.Name -in @(".AL-Go", '.github') } | ForEach-Object {
    Test-Json -jsonFile $_.FullName -baseFolder $baseFolder -repo:($_.BaseName -eq "AL-Go-Settings")
}
```

Comment (freddydk) [reply Actions/AL-Go-TestRepoHelper.ps1]: If I change the code to this, then it will also test files in .AL-Go called oldsettings.json - which could cause a build failure somewhere.
(I actually had a file like that in a local project)

File: Actions/AL-Go-Helper.ps1 (hunk 10 above)
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:349]: done

File: Actions/AL-Go-TestRepoHelper.ps1 (hunk 2 above)
Comment (freddydk) [reply Actions/AL-Go-TestRepoHelper.ps1]: changed the code to a compromise, which shouldn't give any problems in existing repos:

```
    Get-ChildItem -Path $baseFolder -Filter '*.json' -Recurse | ForEach-Object {
        if ($_.DirectoryName -eq '.AL-Go' -and $_.BaseName -eq 'settings') {
            Test-Json -jsonFile $_.FullName -baseFolder $baseFolder
        }
        elseif ($_.DirectoryName -eq '.github' -and $_.BaseName -like '*ettings') {
            Test-Json -jsonFile $_.FullName -baseFolder $baseFolder -repo:($_.BaseName -eq 'AL-Go-Settings')
        }
    }
```

File: Actions/AL-Go-Helper.ps1 [hunk 12]
```diff
@@ -201,38 +214,46 @@ function DownloadAndImportBcContainerHelper {
...
+            $ap = "$ENV:GITHUB_ACTION_PATH".Split([System.IO.Path]::DirectorySeparatorChar)
+            # Action Path under linux is something like: /home/runner/work/_actions/microsoft/AL-Go-Actions/main/WorkflowInitialize
+            # Action Path under Windows is something like: D:\a\_actions\freddydk\AL-Go-Actions\main\WorkflowInitialize
+            # ..../_actions/<owner>/AL-Go-Actions/branch/Action
+            # ..../   -5   /  -4   /     -3      /  -2  /  -1
+            # The code below checks if you are running a preview branch of AL-Go-Actions - then use Preview ContainerHelper
+            # If you are using a private fork of AL-Go-Actions - then use (if exists) a private fork of ContainerHelper as well
+            if ($ap -and $ap.Count -gt 4) {
+                $folder = $ap[$ap.Count-5]
+                $owner = $ap[$ap.Count-4]
```
Comment (mazhelez) [inline Actions/AL-Go-Helper.ps1:235]: Wouldn't it better to use some of the GitHub environment variables, such as `GITHUB_ACTION_REPOSITORY` for the owner/repo name, `GITHUB_REF_NAME` for the branch name, `GITHUB_REPOSITORY_OWNER` for the repo owner?

The way GitHub Actions constructs these paths might change, right?

Reference: https://docs.github.com/en/actions/learn-github-actions/environment-variables

File: Actions/Deliver/Deliver.ps1 [hunk 13]
```diff
@@ -25,13 +25,35 @@ Set-StrictMode -Version 2.0
 $telemetryScope = $null
 $bcContainerHelperPath = $null
 
+function EnsureAzStorageModule() {
+    if (get-command New-AzStorageContext -ErrorAction SilentlyContinue) {
+        Write-Host "Using Az.Storage PowerShell module"
+    }
+    else {
+        $azureStorageModule = Get-Module -name 'Azure.Storage' -ListAvailable | Select-Object -First 1
+        if ($azureStorageModule) {
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/Deliver/Deliver.ps1:34]: AzureRM is pre-installed on all GitHub runners and it does take a fairly long time to install the Az.Storage.

File: Actions/AL-Go-Helper.ps1 (hunk 12 above)
Comment (freddydk) [reply Actions/AL-Go-Helper.ps1]: I needed the branch from the Actions Repo - the github_ref_name doesn't give me that, neither does the github_repository_owner give me the owner of the Actions repo.

But... - the code goes away - I will use the GitHub_action_repository to determine the owner if containerhelperversion is set to private as the only special case.

**Review by mazhelez — APPROVED** [review]
